File: python/src/loxtep/schemas.py
"""
Schemas API (data product schema). get, list, tag_pii_fields.
"""

from __future__ import annotations

from typing import Any, Optional

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient


def _unwrap(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


class SchemasApi:
    """Sync schemas surface."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def get(self, data_product_id: str, *, version: Optional[str] = None) -> dict[str, Any]:
        path = f"/dataproducts/{data_product_id}/schema"
        if version is not None:
            path += f"?version={version}"
        res = self._http.get(path)
        return _unwrap(res)

    def list(self, data_product_id: str) -> dict[str, Any]:
        """List schema versions for a data product. Returns {'items': [...]}."""
        res = self._http.get(f"/dataproducts/{data_product_id}?include_schema=true")
        data = _unwrap(res)
        schema = data.get("schema") if isinstance(data, dict) else None
        items = schema.get("versions", []) if isinstance(schema, dict) else []
        return {"items": items}

    def tag_pii_fields(self, data_product_id: str, fields: list[str]) -> Any:
        """Tag fields as PII on a data product's schema."""
        res = self._http.post(f"/schemas/{data_product_id}/pii", {"fields": fields})
        return _unwrap(res)


class AsyncSchemasApi:
    """Async schemas surface."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def get(self, data_product_id: str, *, version: Optional[str] = None) -> dict[str, Any]:
        path = f"/dataproducts/{data_product_id}/schema"
        if version is not None:
            path += f"?version={version}"
        res = await self._http.get(path)
        return _unwrap(res)

    async def list(self, data_product_id: str) -> dict[str, Any]:
        res = await self._http.get(f"/dataproducts/{data_product_id}?include_schema=true")
        data = _unwrap(res)
        schema = data.get("schema") if isinstance(data, dict) else None
        items = schema.get("versions", []) if isinstance(schema, dict) else []
        return {"items": items}

    async def tag_pii_fields(self, data_product_id: str, fields: list[str]) -> Any:
        res = await self._http.post(f"/schemas/{data_product_id}/pii", {"fields": fields})
        return _unwrap(res)

File: nodejs/src/client/schemas.ts
/**
 * Schemas API (data product schema). get, list. Backend: GET /dataproducts/:id returns schema.
 * list fetches data product with include_schema and returns schema versions.
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type { DataProductSchema, SchemaVersion, SchemaListResponse } from './schemas-types.js';

export function createSchemasApi(http: LoxtepHttpClient): {
  get: (data_product_id: string) => Promise<DataProductSchema | null>;
  list: (data_product_id: string) => Promise<SchemaListResponse>;
  tag_pii_fields: (data_product_id: string, fields: string[]) => Promise<unknown>;
} {
  return {
    async get(data_product_id: string): Promise<DataProductSchema | null> {
      const res = await http.get<{ success?: boolean; data?: { schema?: DataProductSchema } }>(
        `/dataproducts/${encodeURIComponent(data_product_id)}?include_schema=true`
      );
      const data = (res as { data?: { schema?: DataProductSchema } }).data ?? res;
      const schema = (data as { schema?: DataProductSchema }).schema ?? null;
      return schema ?? null;
    },

    async list(data_product_id: string): Promise<SchemaListResponse> {
      const res = await http.get<{
        data?: { schema?: { versions?: SchemaVersion[] } };
        schema?: { versions?: SchemaVersion[] };
      }>(`/dataproducts/${encodeURIComponent(data_product_id)}?include_schema=true`);
      const data = (res as { data?: { schema?: { versions?: SchemaVersion[] } } }).data ?? res;
      const schema = (data as { schema?: { versions?: SchemaVersion[] } }).schema;
      const items = schema?.versions ?? [];
      return { items };
    },

    async tag_pii_fields(data_product_id: string, fields: string[]): Promise<unknown> {
      const res = await http.post<{ success: true; data: unknown }>(
        `/schemas/${encodeURIComponent(data_product_id)}/pii`,
        { fields }
      );
      return res.data;
    },
  };
}
